omegaconf/basecontainer.py

```diff
@@ -189,16 +191,52 @@ def convert(val: Node) -> Any:
 
             return value
 
-        assert isinstance(conf, Container)
+        def get_node_value(key: Union[DictKeyType, int]) -> Any:
+            try:
+                node = conf._get_node(key, throw_on_missing_value=throw_on_missing)
+            except MissingMandatoryValue as e:
+                conf._format_and_raise(key=key, value=None, cause=e)
+            assert isinstance(node, Node)
+            if resolve:
+                try:
+                    node = node._dereference_node()
+                except InterpolationResolutionError as e:
+                    conf._format_and_raise(key=key, value=None, cause=e)
+
+            if isinstance(node, Container):
+                value = BaseContainer._to_content(
+                    node,
+                    resolve=resolve,
+                    throw_on_missing=throw_on_missing,
+                    enum_to_str=enum_to_str,
+                    structured_config_mode=structured_config_mode,
+                )
+            else:
+                value = convert(node)
+            return value
+
         if conf._is_none():
             return None
+        elif conf._is_missing():
+            if throw_on_missing:
+                conf._format_and_raise(
+                    key=None,
+                    value=None,
+                    cause=MissingMandatoryValue("Missing mandatory value"),
+                )
+            else:
+                return MISSING
         elif conf._is_interpolation() and not resolve:
             inter = conf._value()
             assert isinstance(inter, str)
             return inter
-        elif conf._is_missing():
-            return MISSING
-        elif isinstance(conf, DictConfig):
+
+        if resolve:
+            _conf = conf._dereference_node()
+            assert isinstance(_conf, Container)
+            conf = _conf
```

Comment:
I'm curious if this might also be fixing a bug with *valid* top-level interpolations? (it seems that the previous code wasn't calling `_dereference_node()`)

If this is correct, then tests should be added for Dict/ListConfig interpolations.

Side note: `typing.cast()` would make this piece of code cleaner -- but we would lose the assert safety. So far we haven't been using `typing.cast()` so not suggesting you should change it, just an FYI.

Reply:
> I'm curious if this might also be fixing a bug with _valid_ top-level interpolations? (it seems that the previous code wasn't calling `_dereference_node()`)

The previous code was calling `_dereference_node()` inside the `get_node_value` function defined near the top of the `BaseContainer._to_content` method, and inside the `DictConfig._to_object` method (for the case of a `DictConfig` nested inside a structured config, when `SCMode.INSTANTIATE` is being used).

This reminds me of our [previous discussion](https://github.com/omry/omegaconf/pull/730#discussion_r641741533) about about `DictConfig("???")`.
You remarked that such a config could validly appear nested inside a structured config.
The same applies for `DictConfig("${foo}")`: it can appear nested inside a structured config.

That being said, I think that our documented APIs do not currently allow for users to handle `DictConfig("???")` or `DictConfig("${foo}")` directly (as attempting to access them within a structured config would result in `MissingMandatoryValue` in the first case, and in a resolved value or an `InterpolationResolutionError` in the second case). In other words, I think that neither `DictConfig("???")` nor `DictConfig("${foo}")` will ever be passed directly to `OmegaConf.to_container` unless a non-standard/internal APIs is being used.

Currently at the top level of the `BaseContainer._to_content` method there is an `elif conf._is_missing()` branch and an `if resolve` branch (which you highlighted above); these branches are precisely for this situation, where `DictConfig("???")` or `DictConfig("${foo}")` is being passed directly to `OmegaConf.to_container`.

So I think the question is, should we support this even though it's not part of the public API? The alternative would be to fail noisily (with e.g. a `ValueError`) if `DictConfig("???")` or `DictConfig("${foo}")` is passed directly to `OmegaConf.to_container`.

> Side note: `typing.cast()` would make this piece of code cleaner -- but we would lose the assert safety. So far we haven't been using `typing.cast()` so not suggesting you should change it, just an FYI.

Yup, `typing.cast()` came to my mind as I was writing this code :)

Reply:
Just to make my point clear: I think that if we test the case `DictConfig(content="${bar}")` where the `bar` node doesn't exist (current test ID `dictconfig_interp_key_error`), I think we should also test the case where `bar` exists. And I suspect it didn't work before because the code was iterating on the DictConfig keys with `for key in conf.keys():` and `keys()` is empty if `conf` is an interpolation.

An orthogonal question is whether we actually care about this use case -- I'm not 100% sure, but in doubt I'd rather have test cases covering the current behavior, so that we notice if it changes in the future.

Reply:
> Just to make my point clear: I think that if we test the case `DictConfig(content="${bar}")` where the `bar` node doesn't exist (current test ID `dictconfig_interp_key_error`), I think we should also test the case where `bar` exists.

You're right, I'll add some tests.

> And I suspect it didn't work before because the code was iterating on the DictConfig keys with `for key in conf.keys():` and `keys()` is empty if `conf` is an interpolation (but maybe I'm missing something -- I didn't try it).

It actually did work even before commit https://github.com/omry/omegaconf/pull/730/commits/a059a8fc07b733ff3c52b413b0772009ed8c772a was added.
The following works after `git checkout a059a8f~1`, checking out the previous commit:
```
>>> cfg = OmegaConf.create({"a": {"b": 123}, "c": DictConfig("${a}")})
>>> print(OmegaConf.to_container(cfg, resolve=True))
{'a': {'b': 123}, 'c': {'b': 123}}
```
It works because calling `to_container(cfg, resolve=True)` will dereference the interpolation (in `get_node_value`, with `key=="c"`) before calling `_to_content` on that node.

> An orthogonal question is whether we actually care about this use case -- I'm not 100% sure, but in doubt I'd rather have test cases covering the current behavior, so that we notice if it changes in the future.

Sounds good to me.

Reply:
> You're right, I'll add some tests.

I noticed just now that in the `test_to_container` function near the top of `test_to_container.py`, we have the following test cases:
```
        param(
            {"foo": DictConfig(content="${bar}"), "bar": 10},
            {"foo": "${bar}", "bar": 10},
            {"foo": 10, "bar": 10},
            id="dict_inter_dictconfig",
        ),
...
        param(
            {"foo": ListConfig(content="${bar}"), "bar": 10},
            {"foo": "${bar}", "bar": 10},
            {"foo": 10, "bar": 10},
            id="dict_inter_listconfig",
        ),
```
I think that this covers the use-case in question, though it's a bit awkward because the integer `10` is not of the correct type (it should be a dict or list).

Reply:
This is what I was referring to (I just verified that it was indeed not working previously, and your code fixes it):
```
cfg = OmegaConf.create({"foo": DictConfig(content="${bar}"), "bar": {"a": 0}})._get_node("foo")
assert OmegaConf.to_object(cfg) == {"a": 0}
```
==> I think there should be a test for it (both for Dict & ListConfig)

Reply:
Oh, got it :)
Thanks for doing the digging.
I've added the tests in https://github.com/omry/omegaconf/pull/730/commits/f98b1f00e7ae0124395cd3e078d4d03ac5fea2b1.